### DISCUSSION TITLE
Add DataDog for observability

### DISCUSSION THREAD

**Discussion by tohyongcheng** [directus/directus#18006] [Archive]
### Summary

I would like to add Datadog ([dd-trace-js](https://github.com/DataDog/dd-trace-js)) specifically so that I can add some observability into the stack, especially when its used in production for a Backend as a service. 

The way `dd-trace-js` works is that it injects itself to the various libraries and frameworks used (e.g. `express`, `http`, `pino`, `pg`) during the application initialisation of the NodeJS application. The full list of packages they support is at https://github.com/DataDog/dd-trace-js/tree/master/packages. 

The feature would be to enable `dd-trace-js` when an environmental variable such as `ENABLE_DATADOG` is set to true. The good thing is that `dd-trace-js` also works with OpenTelemetry, meaning that we would be able to stream metrics to any metrics provider, or APM solution. 

### Motivation

This gives Directus more production-grade capabilities for it to be used for more enterprise purposes. By having observability, we would be able to monitor and give confidence to developers for those who want to use Directus for production purposes. 

### Detailed Design

We would just need to install the `dd-trace-js` library and add it into the initialisation of the code. See https://docs.datadoghq.com/tracing/trace_collection/dd_libraries/nodejs/?tab=containers.

### Must Have:
- Importing of the `dd-trace-js` library
- Instantialise `dd-trace` upon API loading

### Should Have:
- Environmental variables to support the initialisation of Datadog
- Documentation for enabling Datadog

### Won't Have
- Datadog tracing for Client side (App) yet. 

### Drawbacks

* `dd-trace-js` has history of memory leaks, so you should be careful that it doesn't introduce memory leak issues into Directus before releasing. 

### Alternatives

* There is also Prometheus for monitoring - which is based on exposing endpoints on the application itself. See https://github.com/directus/directus/discussions/17457

### Adoption Strategy

* Through new environmental variables

**Comment by rijkvanzanten** [comment]
I think you can already integrate DataDog by using the `init` hook to directly tie it into the underlying `express` app instance 👍🏻

**Reply by tohyongcheng** [reply]
Oh nice, let me try that.

**Comment by connorwinston** [comment]
> I think you can already integrate DataDog by using the `init` hook to directly tie it into the underlying `express` app instance 👍🏻 

This works 👍, closing this out.
